app/routes/manager/admins.tsx

```diff
@@ -0,0 +1,351 @@
+      response = await addAdminUser({email})
+      if (response.error) {
+        return json({error: response.error}, {status: 404})
+      }
+      return json(response, {status: 201})
+    
+    case 'DELETE':
+      const id = formData.get("id") as string
+      invariant(id, "User Id is required")
+      response = await removeAdminUser({id})
+      if (response.error) {
+        return json({error: response.error}, {status: 400})
+      }
+      return json(response, {status: 200})
+
+    default: {
+      throw new Error("Something went wrong");
+    }
+  }
+};
+
+const GridEditToolbar = (props: gridEditToolbarProps) => {
+  const { setRows, createButtonText } = props;
+
+  const handleAddClick = () => {
+    const id = "new-value";
+    const newName = "";
+    const newEmail = "";
+    setRows((prevRows) => {
+      if (prevRows.find((rowValue) => rowValue.id === "new-value")) {
+        return [...prevRows];
+      }
+      return [
+        ...prevRows,
+        { id, name: newName, email: newEmail, isNew: true },
+      ] as AdminRecord[];
+    });
+  };
+  return (
+    <GridToolbarContainer>
+      <Button
+        variant="contained"
+        color="primary"
+        startIcon={<AddIcon />}
+        onClick={() => handleAddClick()}
+      >
+        {createButtonText}
+      </Button>
+    </GridToolbarContainer>
+  );
+};
+
+export default function  AdminsDataGrid() {
+  const fetcher = useFetcher();
+  const { admins } = useLoaderData();
+  const [error, setError] = useState<string>("");
+  const createButtonText = "Add New Admin";
+  const [rows, setRows] = useState<AdminRecord[]>(() =>
+    admins ? admins.map((item: AdminRecord) => ({
+        id: item.id,
+        name: item.name,
+        email: item.email,
+      }))
+    : []
+  );
+
+  // handle Delete Admin variables
+  const [selectedRowID, setSelectedRowID] = useState("");
+  const [openDeleteModal, setOpenDeleteModal] = useState<boolean>(false);
+
+  useEffect(() =>{
+    //It handles the fetcher error from the response
+    if (fetcher.state === "idle" && fetcher.data) {
+      if (fetcher.data.error){
+        setError(fetcher.data.error);
+      } else {
+        setError("")
+      }
+    }
+  }, [fetcher])
+
+  useEffect(()=> {
+    //It changes the rows shown based on admins
+    setRows(admins.map((item: AdminRecord) => ({
+      id: item.id,
+      name: item.name,
+      email: item.email,
+    })))
+  }, [admins])
+
+  // Set handles for interactions
+  const handleRowEditStart = (event: any) => {
+    event.defaultMuiPrevented = true;
+  };
+
+  const handleRowEditStop = (event: any) => {
+    event.defaultMuiPrevented = true;
+  };
+
+  const handleCellFocusOut = (event: any) => {
+    event.defaultMuiPrevented = true;
+  };
+
+  //Add new Admin
+  const addNewAdminUser = async (values: newAdmin) => {
+    try {
+      const body = {
+        ...values,
+        action: 'POST'
+      }
+      await fetcher.submit(body, { method: "post" });
```

Comment:
have you seen the example at https://remix.run/docs/en/v1/guides/api-routes#call-loaders-outside-of-navigation ?

There they don't use useEffect, but only useFetcher, and instead of doing the `.submit` they use a `.Form` component from the fetcher result.

I haven't been able to try and refactor this PR with that approach 🙈

Reply:
Yep, I saw that, and also tried to implement it, my first approach was to try to refactor the code and use the `<Form />` component from Remix without using the Fetcher, but it's really hard to implement it with the way we use the MUI DataGrid, as the inputs and the submit button are in different columns, I couldn't find a way to group them all in a single Form... The only way I could possibly see is to change the way we add new `ADMINS` by having a Form outside the DataGrid, and not in it as it is today, I don't know if it's worth to do all the refactor just to avoid the `useEffect` though.